Sources/MongoSwift/APM.swift

```diff
@@ -14,11 +14,247 @@ public struct ConnectionId {
+            self.opTime = lastWrite["opTime"] as? ObjectId
+        }
+
+        if let minVersion = isMaster["minWireVersion"] as? Int32 {
+            self.minWireVersion = minVersion
+        }
+
+        if let maxVersion = isMaster["maxWireVersion"] as? Int32 {
+            self.maxWireVersion = maxVersion
+        }
+
+        if let me = isMaster["me"] as? String {
+            self.me = ConnectionId(me)
+        }
+
+        if let hosts = isMaster["hosts"] as? [String] {
+            self.hosts = hosts.map { ConnectionId($0) }
+        }
+
+        if let passives = isMaster["passives"] as? [String] {
+            self.passives = passives.map { ConnectionId($0) }
+        }
+
+        if let arbiters = isMaster["arbiters"] as? [String] {
+            self.arbiters = arbiters.map { ConnectionId($0) }
+        }
+
+        if let tags = isMaster["tags"] as? Document {
+            for (k, v) in tags {
+                self.tags[k] = v as? String
+            }
+        }
+
+        self.setName = isMaster["setName"] as? String
+        self.setVersion = isMaster["setVersion"] as? Int64
+        self.electionId = isMaster["electionId"] as? ObjectId
+
+        if let primary = isMaster["primary"] as? String {
+            self.primary = ConnectionId(primary)
+        }
+
+        self.logicalSessionTimeoutMinutes = isMaster["logicalSessionTimeoutMinutes"] as? Int64
+
+    }
+
+    /// An internal initializer to create a `ServerDescription` from an OpaquePointer to a
+    /// mongoc_server_description_t.
+    internal init(_ description: OpaquePointer) {
+        self.connectionId = ConnectionId(mongoc_server_description_host(description))
+        self.roundTripTime = mongoc_server_description_round_trip_time(description)
+
+        let isMasterData =  UnsafeMutablePointer(mutating: mongoc_server_description_ismaster(description)!)
+        let isMaster = Document(fromPointer: isMasterData)
+        self.parseIsMaster(isMaster)
+
+        let serverType = String(cString: mongoc_server_description_type(description))
+        self.type = ServerType(rawValue: serverType)!
+    }
+}
+
+/// The possible types for a topology. The raw values correspond to the values libmongoc uses. 
+/// (We don't use these strings directly because Swift convention is to use lowercase for enums.)
+public enum TopologyType: String {
+    case single = "Single"
+    case replicaSetNoPrimary = "ReplicaSetNoPrimary"
+    case replicaSetWithPrimary = "ReplicaSetWithPrimary"
+    case sharded = "Sharded"
+    case unknown = "Unknown"
+}
+
+/// A struct describing the state of a MongoDB deployment: its type (standalone, replica set, or sharded), 
+/// which servers are up, what type of servers they are, which is primary, and so on.
+public struct TopologyDescription {
+    /// The type of this topology. 
+    let type: TopologyType
+
+    /// The replica set name. 
+    var setName: String? { return self.servers[0].setName }
```

Comment:
while this is always going to be the same value right now (same goes for `logicalSessionTimeoutMinutes`) and could technically just be a set property, in the future we would want to be dynamically calculating it as servers are added/removed